api/client/gro_client.py

```diff
@@ -7,19 +7,64 @@
 import itertools
 import os
 import sys
+import time
+import json
+
+# Python3 support
+try:
+    # Python3
+    from urllib.parse import urlencode
+except ImportError:
+    # Python2
+    from urllib import urlencode
 
 from api.client import cfg, lib
 from api.client.constants import DATA_SERIES_UNIQUE_TYPES_ID, ENTITY_KEY_TO_TYPE
 from api.client.utils import intersect, zip_selections, dict_unnest
+from api.client.lib import APIError
 
 import pandas
 import unicodecsv
+from tornado import gen
+from tornado.escape import json_decode
+from tornado.httpclient import AsyncHTTPClient, HTTPRequest, HTTPError
+from tornado.ioloop import IOLoop
+from tornado.queues import Queue
 
 
 API_HOST = "api.gro-intelligence.com"
 OUTPUT_FILENAME = "gro_client_output.csv"
 
 
+class BatchError(APIError):
+    """Replicate the APIError interface given a Tornado HTTPError."""
+
+    def __init__(self, response, retry_count, url, params):
+        self.response = response
+        self.retry_count = retry_count
+        self.url = url
+        self.params = params
+        self.status_code = (
+            self.response.code if hasattr(self.response, "code") else None
+        )
+        try:
+            json_content = json_decode(self.response.body)
+            # 'error' should be something like 'Not Found' or 'Bad Request'
+            self.message = json_content.get("error", "")
+            # Some error responses give additional info.
+            # For example, a 400 Bad Request might say "metricId is required"
+            if "message" in json_content:
+                self.message += ": {}".format(json_content["message"])
+        except Exception:
+            # If the error message can't be parsed, fall back to a generic "giving up" message.
+            self.message = "Giving up on {} after {} {}: {}".format(
+                self.url,
+                self.retry_count,
+                "retry" if self.retry_count == 1 else "retries",
+                response,
```

Comment:
minor: extra comma

Reply:
That's actually an intended code style thing: https://black.readthedocs.io/en/stable/the_black_code_style.html#trailing-commas

See https://stackoverflow.com/a/11597911 for some justification of why that's a good thing